impl/src/main/java/org/jboss/weld/bootstrap/events/BuilderInterceptorBean.java

```diff
@@ -0,0 +1,173 @@
+import javax.interceptor.InvocationContext;
+
+import org.jboss.weld.context.WeldCreationalContext;
+import org.jboss.weld.logging.BeanLogger;
+import org.jboss.weld.util.collections.ImmutableSet;
+import org.jboss.weld.util.reflection.Formats;
+
+/**
+ * @author Tomas Remes
+ */
+public class BuilderInterceptorBean implements Interceptor<BuilderInterceptorInstance>, Prioritized, PassivationCapable {
+
+    private final Set<Annotation> bindings;
+
+    private final int priority;
+
+    private final InterceptionType interceptionType;
+
+    private final Function<InvocationContext, Object> interceptorFunction;
+
+    private final BiFunction<InvocationContext, Bean<?>, Object> interceptorMetadataFunction;
+
+    private BuilderInterceptorBean(Set<Annotation> interceptorBindings, InterceptionType type, int priority,
+            Function<InvocationContext, Object> interceptorFunction,
+            BiFunction<InvocationContext, Bean<?>, Object> interceptorMetadataFunction) {
+        this.interceptorFunction = interceptorFunction;
+        this.interceptorMetadataFunction = interceptorMetadataFunction;
+        this.priority = priority;
+        this.interceptionType = type;
+        this.bindings = ImmutableSet.copyOf(interceptorBindings);
+    }
+
+    public BuilderInterceptorBean(Set<Annotation> interceptorBindings, InterceptionType type, int priority,
+            Function<InvocationContext, Object> interceptorFunction) {
+        this(interceptorBindings, type, priority, interceptorFunction, null);
+    }
+
+    public BuilderInterceptorBean(Set<Annotation> interceptorBindings, InterceptionType type, int priority,
+            BiFunction<InvocationContext, Bean<?>, Object> interceptorFunction) {
+        this(interceptorBindings, type, priority, null, interceptorFunction);
+    }
+
+    @Override
+    public Set<Annotation> getInterceptorBindings() {
+        return bindings;
+    }
+
+    @Override
+    public boolean intercepts(InterceptionType type) {
+        return interceptionType.equals(type);
+    }
+
+    @Override
+    public Object intercept(InterceptionType type, BuilderInterceptorInstance builderInterceptorInstance, InvocationContext ctx) throws Exception {
+        if (interceptorMetadataFunction != null) {
+            return interceptorMetadataFunction.apply(ctx, builderInterceptorInstance.getInterceptedBean());
+        } else {
+            return interceptorFunction.apply(ctx);
+        }
+    }
+
+    @Override
+    public Class<?> getBeanClass() {
+        return BuilderInterceptorInstance.class;
+    }
+
+    @Override
+    public Set<InjectionPoint> getInjectionPoints() {
+        return Collections.emptySet();
+    }
+
+    @Override
+    public boolean isNullable() {
+        return false;
+    }
+
+    @Override
+    public Set<Type> getTypes() {
+        HashSet<Type> types = new HashSet<Type>();
+        types.add(BuilderInterceptorInstance.class);
+        types.add(Object.class);
```

Comment:
As I already said - this should be an immutable set cached in a private static final field.